InterMine 3.1.0 -- Add strain to core model.

julie-sullivan opened the pull request:
## Details

See [decisions](https://docs.google.com/document/d/1Xg3E2pUdwt9WKVyHFpIt9M0SBlHXwr2PYllaHsKT7BQ/edit#) for details

Refs #1596

## Testing

We don't have strains to test so will need someone else to do so.

## Checklist

Before your pull request can be approved, be sure to check all boxes:

- [ ] Passing unit test for new or updated code (if applicable)
- [ ] Passes all tests – according to Travis
- [ ] Documentation (if applicable)
- [ ] Single purpose
- [ ] Detailed commit messages
- [ ] Well commented code
- [ ] Checkstyle

sammyjava commented:
I can test this, I'm in a good spot to try it out right now. I'll assume that we're going forward with this because I'll have to majorly hack my loaders to load "strains" differently from how I'm doing it now with multiple organisms. SO ARE YOU SURE ABOUT THIS??? :)

julie-sullivan commented:
@rachellyne What do you think? Safe to go ahead? Or do we want to review as a group again?

julie-sullivan commented:
@rachellyne What do you think? Safe to go ahead? Or do we want to review as a group again?

julie-sullivan commented:
@rachellyne What do you think? Safe to go ahead? Or do we want to review as a group again?

julie-sullivan commented:
@rachellyne What do you think? Safe to go ahead? Or do we want to review as a group again?

julie-sullivan commented:
@rachellyne What do you think? Safe to go ahead? Or do we want to review as a group again?

julie-sullivan commented:
@rachellyne What do you think? Safe to go ahead? Or do we want to review as a group again?

sammyjava commented:
OK, will do.

sammyjava commented:
OK, I've seen you've done some updates to classes and methods, but I'm not sure they make sense. For example, in OrganismRepository:
```
/**
 * Get strains for given taxon ID
 * @param taxonString taxon ID for organism
 * @return taxonId for strain
 */
public String getStrain(String taxonString) {
    return organismsWithStrains.get(taxonString);
}
```
which is populated by
```                            
String[] strains = attributeValue.split(" ");
for (String strain : strains) {
    try {
        or.strains.put(strain, od);
        or.organismsWithStrains.put(taxonIdString, strain);
    } catch (NumberFormatException e) {
        throw new NumberFormatException("taxon ID must be a number");
    }
}
```
But a given `taxonId` can be associated with many strains. It's not a one-to-one correspondence. HashMap requires unique keys, `taxonId` in this case.

A simple example of something that needs to work is `Organism.taxonId=3827` (chickpea) with `Strain.primaryIdentifier=ICC4958` and `Strain.primaryIdentifier=CDCFrontier` in its `strains` collection. That's the simple example (ChickpeaMine) that I'm working on for this update.

I have to make all this work for my loaders, so I was about to dive into updating these core utility objects and methods, but I don't want to stomp on what you may be doing. I'm happy to rewrite this stuff without too much departure, I've got my updates for "variety" support to guide me, although that's a bit different since it was an Organism attribute rather than a collection. Let me know.

julie-sullivan commented:
whoops, no you are right, that should be a one to many relationship. That code is wrong, I will fix!

julie-sullivan commented:
@sammyjava Fixed! Thanks for that.

Also note the organism repository is a static lookup used mainly to get from taxons to abbreviations (UniProt, KEGG, etc. dme --> Drosophila melanogaster).

In this case, strains are misnamed as they are really "subspecies". I added strains to the "organism repository" because one data provider was using a yeast subspecies (taxon 559292) instead of the 4932 taxon ID yeastmine uses and I needed a way to do a quick lookup. 

Maybe I should rename strains here to be subspecies in this class?

sammyjava commented:
No, that would be confusing as f*ck. If it's a taxon ID it's a species. I think that should be a hard and fast rule, whatever the yeasties think. We use taxon ID, not genus/species, as an organism identifier all over the place.

Should I just wait a while on this stuff? Because I'm sure I'll need to edit some things to make my chado (etc.) loader work. I don't want to introduce non-model terminology like subspecies.

sammyjava commented:
OR, I can just fork away and we can hit the merge issues down the line, but then I'm not testing 3.1.0, I'm just forking as I did for "variety" in versions 1 and 2.

sammyjava commented:
OR, just take a break and I'll do pull requests for the changes I make. :)

julie-sullivan commented:
Organism repository is a big, hardcoded lookup from organism/taxon ID to KEGG and UniProt abbreviations. 

The organism repository was not meant to be a lookup from organism to strain, and thus unrelated to our model changes.

julie-sullivan commented:
I used the word "Strain" in that class before we had strain in the core data model. It's not a strain, that's an inaccurate label and should be changed because it's confusing. 

Actually, looking at the code, "strain" is not even used any more. Whoops. I will remove it!

If you need a lookup between organism and strain, you should do a database query. Do you agree?

julie-sullivan commented:
PS. I also use the "strain" label in the UniProt config, which is also incorrect. Might think about renaming that configuration property too.

sammyjava commented:
Yup, gotcha on all that. Actually the only thing I've hit that I think should be in the core model is, for consistency with getOrganismItem, in `BioDBConverter`:
```
    /**
     * Provides the strain item for the given strain name.
     * @param strainName the name of the strain
     * @return the Strain Item
     */
    public Item getStrainItem(String strainName) {
        Item strain = strains.get(strainName);
        if (strain==null) {
            strain = createItem("Strain");
            strain.setAttribute("primaryIdentifier", strainName);
            try {
                store(strain);
            } catch (ObjectStoreException e) {
                throw new RuntimeException("Failed to store strain with name: "+strainName, e);
            }
            strains.put(strainName, strain);
        }
        return strain;
    }
```

julie-sullivan commented:
Yes, I agree, that's a sensible addition!

sammyjava commented:
Travis Checkstyle is the bane of all existence. :-1:

julie-sullivan commented:
* Renamed UniProt's strain label to be more accurate
* Removed OrganismRepository's strain entries because they weren't being used anymore
* Updated version to 3.1.0 so won't work unless you install everything (I kept getting errors switching branches, changing the version was easiest for me)
* Added your updates to BioDBConverter

Thanks so much for your help!!

sammyjava commented:
FYI, I had trouble running postprocess the with flymineVersion in your strain-model branch, and had to update to 
```
systemProp.flymineVersion=4.0.+
```
in the `bio`, `bio/sources` and `bio/postprocess` `gradle.properties` files. It couldn't find the specified version 2.0.+. Not sure why 4.0.0 is in there!

julie-sullivan commented:
@danielabutano Any ideas why that would have happened? I know you published 4.0.0 but the 2.0.0 version should still be there too?

sammyjava commented:
Complete mystery to me as well. FWIW when I fire up a new pull, I wipe everything (including .m2 and .gradle) so I'm working with a totally clean build. If you don't do that, of course, older versions remain and you don't see fun stuff like this. Here's what I run (including my own stuff):
```
#!/bin/sh
rm -rf ~/.gradle/caches/modules-2/files-2.1/org.intermine
rm -rf ~/.m2/repository/org/intermine/*
cd ~/intermine-strains/intermine
./gradlew clean
./gradlew install
cd ../plugin
./gradlew clean
./gradlew install
cd ../bio
./gradlew clean
./gradlew install
cd sources
./gradlew clean
./gradlew install
cd ../postprocess
./gradlew clean
./gradlew install
cd ~/legfed-bio-sources
./gradlew clean
./gradlew install
cd ~/legfed-bio-postprocess
./gradlew clean
./gradlew install
```

julie-sullivan commented:
@danielabutano Ideas? Maybe we should publish 2.0.0 in JCenter too?

sammyjava commented:
Why not just update to 4.0.+ in the `bio`, `bio/sources` and `bio/postprocess` `gradle.properties` files in this branch? Works fine.

sammyjava commented:
Oh, hell, now it's not doing it. gradle versionitis drives me NUTS. No idea, seems to be non-deterministic, irreproducible and otherwise nonsensical. BUT ignore unless I get back to you on this.

sammyjava commented:
So it came back. And I don't get it, seems like the required file is there (although I don't see why we're requiring anything with "flymine" in the name, but that's another story). Here's some dumpage. First, you can see that `gradlew install` in bio/sources grabbed that pom.
```
[shokin@shokin-mines sources]$ ./gradlew install
Download https://oss.jfrog.org/artifactory/oss-snapshot-local/org/intermine/bio-source-flymine-static/2.0.0-RC-SNAPSHOT/bio-source-flymine-static-2.0.0-RC-20180622.133458-5.pom

BUILD SUCCESSFUL in 54s
429 actionable tasks: 195 executed, 234 up-to-date
[shokin@shokin-mines sources]$
```
But then when I try to postprocess beanmine, after a .`gradlew clean`:
```
[shokin@shokin-mines beanmine]$ ./gradlew postprocess --stacktrace

> Configure project :
Using properties file: '/home/shokin/.intermine/beanmine.properties'
cleaning
deleting idresolver cache
deleting datasources.xml
deleting organisms.xml
deleting publications.xml
deleted log files

> Task :dbmodel:initPostProcess
Found 24 post-processes
Performing ALL 24 post-processes

> Task :dbmodel:postProcess
Performing postprocess do-sources.
Performing postprocess legfed-create-gene-flanking-regions.


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':dbmodel:postProcess'.
> Could not resolve all files for configuration ':dbmodel:postProcesses'.
   > Could not find any version that matches org.intermine:bio-source-flymine-static:2.0.+.
     Versions that do not match:
         4.0.0
     Searched in the following locations:
         file:/home/shokin/.m2/repository/org/intermine/bio-source-flymine-static/maven-metadata.xml
         file:/home/shokin/.m2/repository/org/intermine/bio-source-flymine-static/
         https://jcenter.bintray.com/org/intermine/bio-source-flymine-static/maven-metadata.xml
         http://www.ebi.ac.uk/intact/maven/nexus/content/repositories/ebi-repo/org/intermine/bio-source-flymine-static/maven-metadata.xml
         http://www.ebi.ac.uk/intact/maven/nexus/content/repositories/ebi-repo/org/intermine/bio-source-flymine-static/
     Required by:
         project :dbmodel > org.intermine:bio-postprocess-create-references:3.1.0
         project :dbmodel > org.intermine:bio-postprocess-create-chromosome-locations-and-lengths:3.1.0
         project :dbmodel > org.intermine:bio-postprocess-transfer-sequences:3.1.0

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':dbmodel:postProcess'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:63)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:124)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:105)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:99)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:99)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$ArtifactResolveException: Could not resolve all files for configuration ':dbmodel:postProcesses'.
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.rethrowFailure(DefaultConfiguration.java:882)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$1500(DefaultConfiguration.java:114)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.getFiles(DefaultConfiguration.java:856)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.getFiles(DefaultConfiguration.java:382)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration_Decorated.getFiles(Unknown Source)
        at org.gradle.internal.metaobject.BeanDynamicObject$MetaClassAdapter.getProperty(BeanDynamicObject.java:228)
        at org.gradle.internal.metaobject.BeanDynamicObject.tryGetProperty(BeanDynamicObject.java:171)
        at org.gradle.internal.metaobject.CompositeDynamicObject.tryGetProperty(CompositeDynamicObject.java:55)
        at org.gradle.internal.metaobject.AbstractDynamicObject.getProperty(AbstractDynamicObject.java:59)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration_Decorated.getProperty(Unknown Source)
        at org.intermine.plugin.BioSourceProperties.getPostProcessProperties(BioSourceProperties.groovy:62)
        at org.intermine.plugin.BioSourceProperties.getPostProcesserClassName(BioSourceProperties.groovy:32)
        at org.intermine.plugin.BioSourceProperties$getPostProcesserClassName$0.call(Unknown Source)
        at org.intermine.plugin.postprocess.PostProcessPlugin$_apply_closure2$_closure6$_closure7.doCall(PostProcessPlugin.groovy:119)
        at org.intermine.plugin.postprocess.PostProcessPlugin$_apply_closure2$_closure6.doCall(PostProcessPlugin.groovy:67)
        at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:681)
        at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:656)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:122)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:111)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        ... 27 more
Caused by: org.gradle.internal.resolve.ModuleVersionNotFoundException: Could not find any version that matches org.intermine:bio-source-flymine-static:2.0.+.
Versions that do not match:
    4.0.0
Searched in the following locations:
    file:/home/shokin/.m2/repository/org/intermine/bio-source-flymine-static/maven-metadata.xml
    file:/home/shokin/.m2/repository/org/intermine/bio-source-flymine-static/
    https://jcenter.bintray.com/org/intermine/bio-source-flymine-static/maven-metadata.xml
    http://www.ebi.ac.uk/intact/maven/nexus/content/repositories/ebi-repo/org/intermine/bio-source-flymine-static/maven-metadata.xml
    http://www.ebi.ac.uk/intact/maven/nexus/content/repositories/ebi-repo/org/intermine/bio-source-flymine-static/
Required by:
    project :dbmodel > org.intermine:bio-postprocess-create-references:3.1.0
    project :dbmodel > org.intermine:bio-postprocess-create-chromosome-locations-and-lengths:3.1.0
    project :dbmodel > org.intermine:bio-postprocess-transfer-sequences:3.1.0
        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.DynamicVersionResolver.notFound(DynamicVersionResolver.java:105)
        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.DynamicVersionResolver.resolve(DynamicVersionResolver.java:95)
        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.RepositoryChainDependencyToComponentIdResolver.resolve(RepositoryChainDependencyToComponentIdResolver.java:48)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.ComponentResolversChain$DependencyToComponentIdResolverChain.resolve(ComponentResolversChain.java:149)
        at org.gradle.api.internal.artifacts.ivyservice.dependencysubstitution.DependencySubstitutionResolver.resolve(DependencySubstitutionResolver.java:50)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.DependencyGraphBuilder$SelectorState.resolveModuleRevisionId(DependencyGraphBuilder.java:1209)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.DependencyGraphBuilder$EdgeState.resolveModuleRevisionId(DependencyGraphBuilder.java:403)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.DependencyGraphBuilder.performSelectionSerially(DependencyGraphBuilder.java:245)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.DependencyGraphBuilder.resolveEdges(DependencyGraphBuilder.java:210)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.DependencyGraphBuilder.traverseGraph(DependencyGraphBuilder.java:148)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.DependencyGraphBuilder.resolve(DependencyGraphBuilder.java:121)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.DefaultArtifactDependencyResolver.resolve(DefaultArtifactDependencyResolver.java:86)
        at org.gradle.api.internal.artifacts.ivyservice.DefaultConfigurationResolver.resolveGraph(DefaultConfigurationResolver.java:138)
        at org.gradle.api.internal.artifacts.ivyservice.ShortCircuitEmptyConfigurationResolver.resolveGraph(ShortCircuitEmptyConfigurationResolver.java:73)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.resolveGraph(ErrorHandlingConfigurationResolver.java:66)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$4.run(DefaultConfiguration.java:469)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveGraphIfRequired(DefaultConfiguration.java:461)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveToStateOrLater(DefaultConfiguration.java:437)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$1600(DefaultConfiguration.java:114)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.getSelectedArtifacts(DefaultConfiguration.java:865)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.getFiles(DefaultConfiguration.java:853)
        ... 48 more


* Get more help at https://help.gradle.org

BUILD FAILED in 12s
8 actionable tasks: 5 executed, 3 up-to-date
[shokin@shokin-mines beanmine]$ 
```

sammyjava commented:
So you @julie-sullivan found my omission of adding the strain to the Organism.strains collection. But now I'm at a loss as to how to do it in BioDBConverter, where it really needs to be done for any sources that use those methods to handle organisms and, now, strains. The `getOrganismItem` method stores the organism if it's not already stored; but it doesn't yet have any strains added to its strains collection. If I try to store it (again) in `getStrainItem`, it'll ultimately throw an exception when the data is loaded into the production DB:
```
Caused by: : java.lang.RuntimeException: Exception while dataloading - to allow multiple errors, set the property "dataLoader.allowMultipleErrors" to true
Problem while loading item identifier 2_1 because
There are duplicate objects in the source being loaded, multiple items exist with the same item.identifer. Storing again to id 1000002 object from source Organism [annotationVersion=null, assemblyVersion=null, chadoId=null, commonName=null, description=null, genus=null, id=471421, name=null, shortName=null, species=null, taxonId="3827"]
```
Ideally one could un-store an Item and store it again. But I don't think there's such a thing. Here's my update to `getStrainItem` with the offending code commented out, which means I don't fill the Organism.strains collection in `BioDBConverter`. Suggestions???
```
    /**
     * Provides the strain item for the given strain name and associated organism.
     * @param strainName the name of the strain
     * @param taxonId the taxon ID of the organism to which this strain belongs
     * @return the Strain Item
     */
    public Item getStrainItem(String strainName, String taxonId) {
        Item organism = getOrganismItem(taxonId);
        Item strain = strains.get(strainName);
        if (strain == null) {
            strain = createItem("Strain");
            strain.setAttribute("primaryIdentifier", strainName);
            strain.setReference("organism", organism);
            organism.addToCollection("strains", strain);
            try {
                store(strain);
            } catch (ObjectStoreException e) {
                throw new RuntimeException("Failed to store strain with name: " + strainName, e);
            }
            // try {
            //     store(organism);
            // } catch (ObjectStoreException e) {
            //     throw new RuntimeException("Failed to store organism with strain " + strainName + " added to strains.", e);
            // }
            strains.put(strainName, strain);
        }
        return strain;
    }
```

sammyjava commented:
Of course, one distasteful option is to back out my Strain updates to `BioDBConverter` and just know that we cannot use `BioDBConverter.getOrganismItem` if we have a mine that loads strains. Which is punting. And we're not punters.

sammyjava commented:
SO I've spent some time trying to figure out how to get `BioDBConverter` to add strains to the organism.strains collection on the fly, after the organism has been stored in `getOrganismItem`. And I think I came up with the solution, storing the organism intermine ID in a map and using it to add the strain reference. And it looks good in the items database. BUT when the items database is merged into the production DB, I get the following stacktrace. I'll email you the offending `BioDBConverter.java` since I can't upload it here. I think it should work.
```
> Task :dbmodel:loadSingleSource
Loading lis-genomics tgt items into production DB
 [dataLoad] SLF4J: Class path contains multiple SLF4J bindings.
 [dataLoad] SLF4J: Found binding in [jar:file:/home/shokin/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/1.7.7/8095d0b9f7e0a9cd79a663c740e0f8fb31d0e2c8/slf4j-simple-1.7.7.jar!/org/slf4j/impl/StaticLoggerBinder.class]
 [dataLoad] SLF4J: Found binding in [jar:file:/home/shokin/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-jdk14/1.7.10/13675a57601d9106fe82314fe9eebb9b2d9b13c6/slf4j-jdk14-1.7.10.jar!/org/slf4j/impl/StaticLoggerBinder.class]
 [dataLoad] SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
 [dataLoad] SLF4J: Actual binding is of type [org.slf4j.impl.SimpleLoggerFactory]
 [dataLoad] [Task worker for ':' Thread 15] INFO com.zaxxer.hikari.HikariDataSource - db.production - Starting...
 [dataLoad] [Task worker for ':' Thread 15] INFO com.zaxxer.hikari.HikariDataSource - db.production - Start completed.
 [dataLoad] [Task worker for ':' Thread 15] INFO com.zaxxer.hikari.HikariDataSource - db.common-tgt-items - Starting...
 [dataLoad] [Task worker for ':' Thread 15] INFO com.zaxxer.hikari.HikariDataSource - db.common-tgt-items - Start completed.


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':dbmodel:loadSingleSource'.
> java.lang.RuntimeException: java.lang.ClassCastException: org.intermine.objectstore.proxy.ProxyReference cannot be cast to java.util.Set

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':dbmodel:loadSingleSource'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:63)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:124)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:105)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:99)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:99)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: : java.lang.RuntimeException: java.lang.ClassCastException: org.intermine.objectstore.proxy.ProxyReference cannot be cast to java.util.Set
        at org.intermine.dataloader.ObjectStoreDataLoaderTask.execute(ObjectStoreDataLoaderTask.java:132)
        at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:293)
        at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
        at org.intermine.plugin.integrate.IntegrateUtils$_closure10.doCall(IntegrateUtils.groovy:263)
        at org.intermine.plugin.integrate.IntegratePlugin$_apply_closure9$_closure24.doCall(IntegratePlugin.groovy:222)
        at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:681)
        at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:656)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:122)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:111)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        ... 27 more
Caused by: java.lang.RuntimeException: java.lang.ClassCastException: org.intermine.objectstore.proxy.ProxyReference cannot be cast to java.util.Set
        at org.intermine.dataconversion.ItemToObjectTranslator.translateFromDbObject(ItemToObjectTranslator.java:418)
        at org.intermine.objectstore.translating.ObjectStoreTranslatingImpl.execute(ObjectStoreTranslatingImpl.java:171)
        at org.intermine.objectstore.fastcollections.ObjectStoreFastCollectionsForTranslatorImpl.execute(ObjectStoreFastCollectionsForTranslatorImpl.java:193)
        at org.intermine.dataloader.BatchingFetcher$NoseyObjectStore.execute(BatchingFetcher.java:580)
        at org.intermine.objectstore.query.ResultsBatches.fetchBatchFromObjectStore(ResultsBatches.java:193)
        at org.intermine.objectstore.query.PrefetchManager.doRequest(PrefetchManager.java:241)
        at org.intermine.objectstore.query.PrefetchManager.doRequest(PrefetchManager.java:146)
        at org.intermine.objectstore.query.ResultsBatches.getBatch(ResultsBatches.java:164)
        at org.intermine.objectstore.query.ResultsBatches.getRowsFromBatch(ResultsBatches.java:139)
        at org.intermine.objectstore.query.Results.range(Results.java:197)
        at org.intermine.objectstore.query.SingletonResults.range(SingletonResults.java:66)
        at org.intermine.objectstore.query.Results.get(Results.java:226)
        at org.intermine.objectstore.query.Results$Iter.hasNext(Results.java:378)
        at org.intermine.dataloader.ObjectStoreDataLoader.process(ObjectStoreDataLoader.java:139)
        at org.intermine.dataloader.ObjectStoreDataLoader.process(ObjectStoreDataLoader.java:60)
        at org.intermine.dataloader.ObjectStoreDataLoaderTask.execute(ObjectStoreDataLoaderTask.java:127)
        ... 40 more
Caused by: java.lang.ClassCastException: org.intermine.objectstore.proxy.ProxyReference cannot be cast to java.util.Set
        at org.intermine.model.bio.OrganismShadow.setFieldValue(OrganismShadow.java:185)
        at org.intermine.dataconversion.ItemToObjectTranslator.translateFromDbObject(ItemToObjectTranslator.java:367)
        ... 55 more


* Get more help at https://help.gradle.org

BUILD FAILED in 3m 25s
```

sammyjava commented:
@sergiocontrino check out the stacktrace above from using DataConverter.storeReference. I think it should have worked. https://github.com/intermine/intermine/pull/1926#issuecomment-439222470

sergiocontrino commented:
hi @sammyjava ,
i could not find anything wrong with your code, i reserve to try it later.
i found an example similar to your case that should still work (where experimental factors collections are added to a list of submissions). let me know if it is not useful.

https://github.com/modENCODE-DCC/modmine/blob/dev/bio/sources/chado-db/main/src/org/intermine/bio/dataconversion/ModEncodeMetaDataProcessor.java#L3873-L3900

sammyjava commented:
Thanks, Sergio, I'll look at that and keep poking at it, maybe I'll find a "sweet spot" in the API that works. :)

julie-sullivan commented:
1. We're happy to collect the organism (and strain) items and store them at the end. There shouldn't be many organisms, not so much as to cause memory issues anyway!

2. Sam has tested, Contrino has tested, and I've run builds on FlyMine and HumanMine and everything looks good. Of course the community reviewed as well.

3. all tests pass!
